### PR TITLE
fix(s3): more accurate expiration judgment when starting the cluster

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/object_store/src/object/s3.rs [hunk 1]
```diff
@@ -716,11 +716,28 @@ impl S3ObjectStore {
...
         if let Ok(config) = &get_config_result {
             for rule in config.rules().unwrap_or_default() {
-                if rule.expiration().is_some() {
-                    is_expiration_configured = true;
+                // Check if the filter is not set or the prifix in the filter is data directory in RisingWave,
+                // and if the expiration status of the rule is "Enabled".
+                // If both conditions are met, it is considered that there is a risk of data deletion.
+                match rule.filter().as_ref() {
+                    Some(&LifecycleRuleFilter::Prefix(prefix)) => {
+                        if let Some(ExpirationStatus::Enabled) = rule.status && prefix == data_directory{
```
Comment (zwang28) [inline src/object_store/src/object/s3.rs:727]: better to check whether data directory contains prefix.

**Review by zwang28 — APPROVED** [review]
rest lgtm

File: src/object_store/src/object/s3.rs (hunk 1 above)
Comment (hzxa21) [reply src/object_store/src/object/s3.rs:727]: +1

Comment (wcy-fdu) [reply src/object_store/src/object/s3.rs:727]: Nice catch.

File: src/object_store/src/object/s3.rs [hunk 2]
```diff
@@ -724,7 +724,7 @@ impl S3ObjectStore {
                 // If both conditions are met, it is considered that there is a risk of data deletion.
                 match rule.filter().as_ref() {
                     Some(&LifecycleRuleFilter::Prefix(prefix)) => {
-                        if let Some(ExpirationStatus::Enabled) = rule.status && prefix == data_directory{
+                        if let Some(ExpirationStatus::Enabled) = rule.status && data_directory.contains(prefix){
```
Comment (hzxa21) [inline src/object_store/src/object/s3.rs:727]: hmm... Shouldn't this be `prefix.starts_with(data_directory)`?

Comment (wcy-fdu) [reply src/object_store/src/object/s3.rs:727]: I think it should be `data_directory.start_with(prefix)`, eg, `prefix=hummock`, `data_directory = hummock_001`, and hummock_001 has rule.

Comment (hzxa21) [reply src/object_store/src/object/s3.rs:727]: I get it wrong, it should be `data_directory.start_with(prefix)`.

File: src/object_store/src/object/s3.rs [hunk 3]
```diff
@@ -716,11 +716,28 @@ impl S3ObjectStore {
...
+                        }
+                    }
+                    None => {
+                        if let Some(ExpirationStatus::Enabled) = rule.status {
+                            is_expiration_configured = true;
+                        }
+                    }
+                    _ => {
+
+                    }
```
Comment (hzxa21) [inline src/object_store/src/object/s3.rs:738]: We should wrap this check under `if rule.expiration.is_some()`

**Pull request merged**